WIP: Add integration testing support

simojo opened the pull request:
**This PR is a feature that fixes #197.** This PR provides support for integration testing of sample/starter repositories under [Gator Educator](https://github.com/GatorEducator/).

Simply put, this feature will halt the build process if `gradle grade` shows any inconsistency (i.e. it passes starter repositories and fails solutions). _For a technical description, see code._

@gkapfham please note that there exist gaps to fill, tests to write, and that this code has not yet been tested using `gradle`. Due to what hinders us to meet in person, I hope this PR will provide enough context for the time being.

simojo committed:
method to fetch repos for GatorEd

simojo committed:
helper functions for integration testing

Michionlion commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Complexity increasing per file
==============================
- scripts/ensure.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GatorEducator/gatorgrader/pullRequest?prid=6342427&bid=20600891)

simojo committed:
Modify formatting

Michionlion commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Complexity increasing per file
==============================
- scripts/ensure.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GatorEducator/gatorgrader/pullRequest?prid=6342427&bid=20600891)

simojo committed:
add docstrings for all funcs

Michionlion commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- scripts/ensure.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GatorEducator/gatorgrader/pullRequest?prid=6342427&bid=20600891)

gkapfham commented:
Hello @simojo can you please provide documentation about how to run this code?

simojo committed:
added more documentation & fixed bug with directory scaffolding

simojo committed:
keeping comments up to date

simojo committed:
using black to reformat

simojo committed:
fix documentation build errors

simojo committed:
fixing more documentation build errors

simojo committed:
fix: disposable variable naming convention

simojo committed:
deleted jupyter specific files that were wreaking havoc on flake8

simojo committed:
checks status code of cloning for better syntax practice

simojo commented:
@gkapfham I believe the only build errors left are ones pertaining to the ruby package manager and other files I did not touch!

simojo committed:
doc: we don't need async

simojo commented:
for python proc management

https://docs.python.org/2/library/os.html#process-management

corlettim commented:
Hi @simojo ! I am just reaching out to see if you need any help with working on this PR during this semester. If so, please let me or anyone else on the GatorGrader team! We would be happy to help. I see that your PR is out of date with master so if you are able to update it, it would let us see if the errors still exist and we can go from there.
We also have weekly meetings on Wednesday and if you would like to join, please let me know and we can set you up with the info. Thanks!

simojo commented:
Hi @corlettim! Thanks for the message. I'll fast forward the branch to be even with master, thanks for reminding me.

simojo committed:
minor error handling under main function

simojo commented:
Alright, @corlettim, it's rebased on top of the upstream master! Feel free to review it; the functionality is there, but there may be minor wrinkles in the semantics. It essentially clones down every solution and starter repo and grades both of them, ensuring that the starters fail and the solutions pass.

corlettim commented:
> Alright, @corlettim, it's rebased on top of the upstream master! Feel free to review it; the functionality is there, but there may be minor wrinkles in the semantics. It essentially clones down every solution and starter repo and grades both of them, ensuring that the starters fail and the solutions pass.

Awesome!

corlettim commented:
Hey @simojo ! We are made some changes in the master branch by updating the dependencies so please make sure to resolve the conflicts you are currently having with your PR. Also, when you get the chance, if you could give us an update on what your next steps are with this PR so the software engineering team can help in anyway possible.

simojo committed:
Merge branch 'master' into feature/integration-testing

simojo committed:
dropping colors

It will be better to leave self implemented coloring out because this will run on windows, mac, and possibly zsh; not only bash.

simojo commented:
Ideally, my next steps would be merge it. Apparently there is some linting error with invoking subprocesses without a shell; does your team have a specific way you approach invoking shell commands via python? I don't think security is really that big of an issue here, but the linter says otherwise.

simojo commented:
Are there any plans to merge this any time soon?

gkapfham commented:
This PR is being closed because it reflects the use of an old build system.
